ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp

```diff
@@ -0,0 +1,123 @@
+                        if(host_reshape) {
+                            return  host_row_major_reshape(tensor, shape);
+                        }
+                        else {
+                            auto tensor_4d = unsqueeze_to_4D(tensor);
+                            const auto shape_4d = shape.to_rank<4>();
+                            auto rm_tensor = ttnn::to_layout(tensor_4d, ttnn::ROW_MAJOR_LAYOUT, std::nullopt, std::nullopt, (Device *)nullptr);
+                            auto unsqueezed_tensor = ttnn::reshape_on_device(rm_tensor, shape_4d[0], shape_4d[1], shape_4d[2], shape_4d[3], tensor.memory_config());
+                            auto squeezed_tensor = squeeze_from_4D(unsqueezed_tensor, original_rank);
+                            return ttnn::to_layout(squeezed_tensor, layout, std::nullopt, std::nullopt, (Device *)nullptr);
+                        }
+                    }
+                } else {
+                    return tensor.reshape(shape.value);
+                }
+            } else if (tensor_shape.rank() >= 2 and shape.rank() >= 2) {
+                // Handle the case when the tensor is not contiguous but the last two dimensions are the same and so reshape
+                // is possible
+                if (tensor_shape[-1] == shape[-1] and tensor_shape[-2] == shape[-2] and
+                    tensor_shape.with_tile_padding()[-1] == shape.with_tile_padding()[-1] and
+                    tensor_shape.with_tile_padding()[-2] == shape.with_tile_padding()[-2]) {
+                    return tensor.reshape(shape.value);
+                }
+            }
+            return host_row_major_reshape(tensor, shape);
+    }
+
+
+}
+
+
+ttnn::Tensor ReshapeViewOperation::invoke(
+    const ttnn::Tensor& tensor,
+    const ttnn::Shape& shape
+    ) {
+
+    auto tensor_shape = tensor.get_shape();
+    if (tensor_shape == shape) {
+        return {tensor};
+    }
+
+    const auto layout = tensor.get_layout();
+
+    if (layout == ttnn::Layout::ROW_MAJOR) {
+        return detail::row_major_reshape(tensor, shape);
+    } else {
+        const auto new_shape_with_tile_padding = shape.with_tile_padding();
+        const auto new_height = new_shape_with_tile_padding[-2];
+        const auto new_width = new_shape_with_tile_padding[-1];
+
+        const auto is_tile_multiple = (new_height % ttnn::TILE_SIZE == 0 && new_width % ttnn::TILE_SIZE == 0);
+        if (not is_tile_multiple) {
+            return detail::row_major_reshape(tensor, shape);
+        }
+
+        if (ttnn::has_storage_type_of(tensor, ttnn::StorageType::DEVICE)) {
+            if (tensor_shape.with_tile_padding()[-1] == new_width) {
+                return tensor.reshape(shape.value);
+            }
+        } else {
+            return tensor.reshape(shape.value);
+        }
+    }
+    return detail::row_major_reshape(tensor, shape);
+
+
+}
```

Comment:
1. In this function there are 3 calls to `row_major_reshape` alone. I count 6 calls to `tensor.reshape(shape.value)` across 3 functions. Besides that there is one call to `..., data_transfer_to_device` and one `to_layout`

Could the code be fixed up to something look like the following (I don't have preferences for if vs else if and whether helper functions are used to evaluate the booleans)?
```
    if (tensor_shape == shape) {
        return {tensor};
    }
    if (do_to_layout) {
        ...
        return to_layout...;
   }
    if (do_device) {
        ...
        return data_transfer_to_device...;
   }
   return tensor.reshape(shape.value)
```

2. The original code had a path were reshape was not handled. With this change, will all possible shapes be handled? Or can there be a potential downstream assert/fatal that would be triggered?

Reply:
Agree

Reply:
Thanks @bbradelTT I cleaned it up, the only case that I know we can't handle at the moment is more than 4D tensors, for which I added a FATAL